### PR TITLE
[LibOS] Return -EISDIR on opening a directory with O_CREAT flag

### LINKED ISSUE #2550: Opening dir should fail with EISDIR but gets opened successfully
## Description of the problem
1. Opening device special file with O_RDONLY or O_WRONLY or O_RDWR permission fails
2. Opening a symlink dir with O_RDONLY | O_CREAT should fails with error EISDIR but got SUCCESS
3. Opening a dir with O_RDONLY | O_CREAT should also fail with EISDIR error but got success

## Steps to reproduce
Execute attached files
open11_setup in linux environment and open11_run with graphene-direct
$ ./open11_setup
$ graphene-direct ./open11_run

## Expected results
```
TINFO: Timeout per run is 0h 05m 00s
TPASS: Open regular file O_RDONLY
TPASS: Open regular file O_WRONLY
TPASS: Open regular file O_RDWR
TPASS: Open regular file O_RDWR | O_SYNC
TPASS: Open regular file O_RDWR | O_TRUNC
TPASS: Open dir O_RDONLY
TPASS: Open dir O_RDWR, expect EISDIR
TPASS: Open regular file O_DIRECTORY, expect ENOTDIR
TPASS: Open hard link file O_RDONLY
TPASS: Open hard link file O_WRONLY
TPASS: Open hard link file O_RDWR
TPASS: Open sym link file O_RDONLY
TPASS: Open sym link file O_WRONLY
TPASS: Open sym link file O_RDWR
TPASS: Open sym link dir O_RDONLY
TPASS: Open sym link dir O_WRONLY, expect EISDIR
TPASS: Open sym link dir O_RDWR, expect EISDIR
TPASS: Open device special file O_RDONLY
TPASS: Open device special file O_WRONLY
TPASS: Open device special file O_RDWR
TPASS: Open non-existing regular file in existing dir
TPASS: Open link file O_RDONLY | O_CREAT
TPASS: Open symlink file O_RDONLY | O_CREAT
TPASS: Open regular file O_RDONLY | O_CREAT
TPASS: Open symlink dir O_RDONLY | O_CREAT, expect EISDIR
TPASS: Open dir O_RDONLY | O_CREAT, expect EISDIR
TPASS: Open regular file O_RDONLY | O_TRUNC, behaviour is undefined but should not oops or hang
TPASS: Open regular file(non-empty) O_RDONLY | O_TRUNC, behaviour is undefined but should not oops or hang

Summary:
passed   28
failed   0
skipped  0
warnings 0
```

## Actual results
```
TINFO: Timeout per run is 0h 05m 00s
TPASS: Open regular file O_RDONLY
TPASS: Open regular file O_WRONLY
TPASS: Open regular file O_RDWR
TPASS: Open regular file O_RDWR | O_SYNC
TPASS: Open regular file O_RDWR | O_TRUNC
TPASS: Open dir O_RDONLY
TPASS: Open dir O_RDWR, expect EISDIR
TPASS: Open regular file O_DIRECTORY, expect ENOTDIR
TPASS: Open hard link file O_RDONLY
TPASS: Open hard link file O_WRONLY
TPASS: Open hard link file O_RDWR
TPASS: Open symlink file O_RDONLY
TPASS: Open symlink file O_WRONLY
TPASS: Open symlink file O_RDWR
TPASS: Open symlink dir O_RDONLY
TPASS: Open symlink dir O_WRONLY, expect EISDIR
TPASS: Open sym link dir O_RDWR, expect EISDIR

> TFAIL: Open device special file O_RDONLY - expected SUCCESS: EACCES (13)
> TFAIL: Open device special file O_WRONLY - expected SUCCESS: EACCES (13)
> TFAIL: Open device special file O_RDWR - expected SUCCESS: EACCES (13)

TPASS: Open non-existing regular file in existing dir
TPASS: Open link file O_RDONLY | O_CREAT
TPASS: Open symlink file O_RDONLY | O_CREAT
TPASS: Open regular file O_RDONLY | O_CREAT

> TFAIL: Open symlink dir O_RDONLY | O_CREAT, expect EISDIR - expected EISDIR: SUCCESS (0)
> TFAIL: Open dir O_RDONLY | O_CREAT, expect EISDIR - expected EISDIR: SUCCESS (0)

TPASS: Open regular file O_RDONLY | O_TRUNC, behaviour is undefined but should not oops or hang
TPASS: Open regular file(non-empty) O_RDONLY | O_TRUNC, behaviour is undefined but should not oops or hang

Summary:
passed   23
failed   5
skipped  0
warnings 0
```
[open11.zip](https://github.com/oscarlab/graphene/files/6815909/open11.zip)

### PR DESCRIPTION
**Pull request by sahason**
## Description of the changes 

Opening a directory with `O_CREAT` should fail with `-EISDIR` instead of success.
Fixes #2550.

## How to test this PR? 

Follow the steps to reproduce mentioned in issue #2550.

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)

### CODE REVIEW

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

**Comment by mkow** [comment]
Jenkins, test this please

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @sahason)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-MiWAWWM8bCeA3sU1Uwu:bn0d1h6) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
> ```C
> 
>     if (dent->type == S_IFDIR) {
>         if (flags & O_WRONLY || flags & O_RDWR || flags & O_CREAT) {
> ```

This is not enough because `open_namei()` is also called by `shim_do_mkdirat()`, see: https://github.com/oscarlab/graphene/blob/6ee2c0b15b4f507c15d15cac733c091080246b04/LibOS/shim/src/sys/shim_file.c#L98

This is way our Jenkins CI failed. Here is what Jenkins complains about:
```
Failed: test_libos.TC_30_Syscall.test_023_readdir
mkdir failed: Is a directory
```

I suggest to distinguish between `man 2 open` and `man 2 mkdir` in here as follows:
```
    if (dent->type == S_IFDIR) {
        if (flags & O_WRONLY || flags & O_RDWR || ((flags & O_CREAT) && !(flags & O_DIRECTORY)) {
            ... same ...
```

In other words, we will return `-EISDIR` if the user syscall requested creation (`O_CREAT`) on a regular file (`!O_DIRECTORY`) but the found path is actually a directory.

@pwmarcz Does this sound reasonable to you?

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-MiWUQceEHhT9Hj3MuqE:blkbsdc) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This is not enough because `open_namei()` is also called by `shim_do_mkdirat()`, see: https://github.com/oscarlab/graphene/blob/6ee2c0b15b4f507c15d15cac733c091080246b04/LibOS/shim/src/sys/shim_file.c#L98

This is way our Jenkins CI failed. Here is what Jenkins complains about:
```
Failed: test_libos.TC_30_Syscall.test_023_readdir
mkdir failed: Is a directory
```

I suggest to distinguish between `man 2 open` and `man 2 mkdir` in here as follows:
```
    if (dent->type == S_IFDIR) {
        if (flags & O_WRONLY || flags & O_RDWR || ((flags & O_CREAT) && !(flags & O_DIRECTORY)) {
            ... same ...
```

In other words, we will return `-EISDIR` if the user syscall requested creation (`O_CREAT`) on a regular file (`!O_DIRECTORY`) but the found path is actually a directory.

@pwmarcz Does this sound reasonable to you?
</blockquote></details>

Almost. Your condition will also be triggered on `open(O_CREAT | O_EXCL)`, and in that case we want to return `-EEXIST`, not `-EISDIR`. So you need to check for `!(flags & O_EXCL)` as well.

**Review by sahason — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-Miai1gJA_MVY3bKVQ16:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Almost. Your condition will also be triggered on `open(O_CREAT | O_EXCL)`, and in that case we want to return `-EEXIST`, not `-EISDIR`. So you need to check for `!(flags & O_EXCL)` as well.
</blockquote></details>

Done.

**Comment by mkow** [comment]
Jenkins, retest this please

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @mkow and @pwmarcz)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-MieZtx_BY49kGcrfTNd:b-pnv64e) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
<details><summary><i>Previously, sahason wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think Pawel meant this:
```
    if (dent->type == S_IFDIR) {
        if (flags & O_WRONLY || flags & O_RDWR ||
                ((flags & O_CREAT) && !(flags & O_DIRECTORY) && !(flags & O_EXCL)) {
            ... same ...
```

@pwmarcz Is that what you wanted?

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-MifgIkC4kk_QghSujuh:b-z8piuj) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I think Pawel meant this:
```
    if (dent->type == S_IFDIR) {
        if (flags & O_WRONLY || flags & O_RDWR ||
                ((flags & O_CREAT) && !(flags & O_DIRECTORY) && !(flags & O_EXCL)) {
            ... same ...
```

@pwmarcz Is that what you wanted?
</blockquote></details>

Yes. The current condition will still be true for `O_CREAT | O_EXCL` flags, so it's wrong. @dimakuv's proposed fix looks right.

**Review by sahason — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[LibOS/shim/src/fs/shim_namei.c, line 525 at r1](https://reviewable.io/reviews/oscarlab/graphene/2671#-MiWAWWM8bCeA3sU1Uwt:-MitLzxt3T5LhnhfhWY5:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/1ee9f081ecf882833c0e9eb4dd8cdaf9322969ef/LibOS/shim/src/fs/shim_namei.c#L525)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Yes. The current condition will still be true for `O_CREAT | O_EXCL` flags, so it's wrong. @dimakuv's proposed fix looks right.
</blockquote></details>

Done.

**Review by dimakuv — DISMISSED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by pwmarcz — DISMISSED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by pwmarcz — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2671)* status: :shipit: complete! all files reviewed, all discussions resolved
